kmod/core/core.c

```diff
@@ -890,33 +892,63 @@ static int kpatch_module_notify(struct notifier_block *nb, unsigned long action,
 	if (!found)
 		goto out;
 
-	ret = kpatch_link_object(kpmod, object);
-	if (ret)
-		goto out;
+	if (action == MODULE_STATE_COMING) {
 
-	BUG_ON(!object->mod);
+		ret = kpatch_link_object(kpmod, object);
+		if (ret)
+			goto coming_out;
 
-	pr_notice("patching newly loaded module '%s'\n", object->name);
+		BUG_ON(!object->mod);
 
-	/* run user-defined pre-patch callback */
-	ret = pre_patch_callback(object);
-	if (ret) {
-		pr_err("pre-patch callback failed!\n");
-		goto out;	/* and WARN */
-	}
+		pr_notice("patching newly loaded module '%s'\n", object->name);
 
-	/* add to the global func hash */
-	list_for_each_entry(func, &object->funcs, list)
-		hash_add_rcu(kpatch_func_hash, &func->node, func->old_addr);
+		/* run user-defined pre-patch callback */
+		ret = pre_patch_callback(object);
+		if (ret) {
+			pr_err("pre-patch callback failed!\n");
+			goto coming_out;	/* and WARN */
+		}
+
+		/* add to the global func hash */
+		list_for_each_entry(func, &object->funcs, list)
+			hash_add_rcu(kpatch_func_hash, &func->node,
+				     func->old_addr);
+
+		/* run user-defined post-patch callback */
+		post_patch_callback(object);
+coming_out:
+		/* no way to stop the module load on error */
+		WARN(ret, "error (%d) patching newly loaded module '%s'\n", ret,
+		     object->name);
+
+	} else if (action == MODULE_STATE_GOING) {
+
+		/*
+		 * This case handles the rare instance that a loading
+		 * module has failed to initialize.  We need to invoke
+		 * any unpatch callbacks, remove its functions from
+		 * the kpatch_func_hash and unlink the object.
+		 */
+
+		/* run user-defined pre-unpatch callback */
+		pre_unpatch_callback(object);
+
+		/* remove from the global func hash */
+		list_for_each_entry(func, &object->funcs, list)
+			hash_del_rcu(&func->node);
+
+		/* run user-defined pre-unpatch callback */
+		post_unpatch_callback(object);
 
-	/* run user-defined post-patch callback */
-	post_patch_callback(object);
+		/*
+		 * can't unlink an object that has been removed from the
+		 * global hash, so just clear its mod pointer.
+		 */
+		object->mod = NULL;
```

Comment:
To be perfectly symmetric, I wanted to call `kpatch_link_object()` here, but that fails with "can't remove ftrace filter at address".  Am I missing something?  Since the module isn't going to be loaded anyway, clearing the `mod` pointer was good enough to trip up any future `kpatch_object_linked()` checks.

Reply:
Hmm, if I move this teardown code up into the module coming case (right after it runs the post-patch callback), it all runs w/o incident.  I suspect you're right that something external to kpatch is getting in the way here.  I'm looking through the `load_module()` and `do_init_module()` code now, though nothing is jumping out yet.

Reply:
Bingo, here it is: `ftrace_module_notify_exit() -> ftrace_release_mod(mod)`   (And renamed in newer RHEL7 kernels as `ftrace_module_notify()`.  Looking that the `module_notify_list` in crash, I can see that it's before the kpatch module notifier.

Reply:
Ah, then we may end up having to register *two* notifiers: the coming notifier with priority INT_MIN so it's called last, and the going notifier with priority INT_MAX, so it's called first.